crates/ra_project_model/src/lib.rs

```diff
@@ -1,121 +1,221 @@
 mod cargo_workspace;
+mod json_project;
 mod sysroot;
 
-use std::path::{Path, PathBuf};
+use std::{
+    fs::File,
+    io::BufReader,
+    path::{Path, PathBuf},
+};
 
 use failure::bail;
 use rustc_hash::FxHashMap;
 
 use ra_db::{CrateGraph, FileId, Edition};
 
+use serde_json::from_reader;
+
 pub use crate::{
     cargo_workspace::{CargoWorkspace, Package, Target, TargetKind},
+    json_project::JsonProject,
     sysroot::Sysroot,
 };
 
 // TODO use proper error enum
 pub type Result<T> = ::std::result::Result<T, ::failure::Error>;
 
 #[derive(Debug, Clone)]
-pub struct ProjectWorkspace {
-    pub cargo: CargoWorkspace,
-    pub sysroot: Sysroot,
+pub enum ProjectWorkspace {
+    /// Project workspace was discovered by running `cargo metadata` and `rustc --print sysroot`.
+    Cargo { cargo: CargoWorkspace, sysroot: Sysroot },
+    /// Project workspace was manually specified using a `rust-project.json` file.
+    Json { project: JsonProject },
 }
 
 impl ProjectWorkspace {
     pub fn discover(path: &Path) -> Result<ProjectWorkspace> {
-        let cargo_toml = find_cargo_toml(path)?;
-        let cargo = CargoWorkspace::from_cargo_metadata(&cargo_toml)?;
-        let sysroot = Sysroot::discover(&cargo_toml)?;
-        let res = ProjectWorkspace { cargo, sysroot };
-        Ok(res)
+        match find_rust_project_json(path) {
+            Some(json_path) => {
+                let file = File::open(json_path)?;
+                let reader = BufReader::new(file);
+                Ok(ProjectWorkspace::Json { project: from_reader(reader)? })
+            }
+            None => {
+                let cargo_toml = find_cargo_toml(path)?;
+                Ok(ProjectWorkspace::Cargo {
+                    cargo: CargoWorkspace::from_cargo_metadata(&cargo_toml)?,
+                    sysroot: Sysroot::discover(&cargo_toml)?,
+                })
+            }
+        }
     }
 
-    pub fn to_crate_graph(&self, load: &mut dyn FnMut(&Path) -> Option<FileId>) -> CrateGraph {
-        let mut crate_graph = CrateGraph::default();
-        let mut sysroot_crates = FxHashMap::default();
-        for krate in self.sysroot.crates() {
-            if let Some(file_id) = load(krate.root(&self.sysroot)) {
-                sysroot_crates
-                    .insert(krate, crate_graph.add_crate_root(file_id, Edition::Edition2015));
+    pub fn add_roots(&self, roots: &mut Vec<PathBuf>) {
```

Comment:
The name is somewhat unfortunate: it sounds like `fn add_roots(&mut self)`. Let's change signature to `-> Vec<PathBuf>` or `-> impl Iterator<&Path>`?

Reply:
Fixed.